src/pdm_low_level_l476rg.cpp

```diff
@@ -0,0 +1,388 @@
+extern "C" {
+#endif
+
+uint32_t SAI1_client = 0;
+SAI_HandleTypeDef hsai_BlockB1;
+DMA_HandleTypeDef hdma_sai1_b;
+uint16_t SAI_InternalBuffer[((AUDIO_IN_FREQ / 8) * MAX_AUDIO_IN_CHANNEL_NBR_PER_IF * N_MS_PER_INTERRUPT)];
+uint32_t SAI_InternalBufferSize = (((AUDIO_IN_FREQ / 8) * MAX_AUDIO_IN_CHANNEL_NBR_PER_IF *N_MS_PER_INTERRUPT) * 2);
+
+static void MX_DMA_Init(void);
+static int MX_SAI1_Init(void);
+
+/* Low Level Functions implementation */
+
+/**
+ *  @brief PDM initialization
+ *  @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_init(void)
+{
+  int ret = PDM_OK;
+  MX_DMA_Init();
+  ret = MX_SAI1_Init();
+  return ret;
+}
+
+/**
+ *  @brief PDM deinitialization
+  * @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_deinit(void)
+{
+  if (HAL_SAI_DeInit(&hsai_BlockB1) != HAL_OK) {
+    return PDM_ERROR;
+  }
+
+  return PDM_OK;
+}
+
+/**
+ *  @brief Start recording
+  * @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_record(void)
+{
+  if (HAL_SAI_Receive_DMA(&hsai_BlockB1, (uint8_t *)SAI_InternalBuffer, (uint16_t)SAI_InternalBufferSize / 2U) != HAL_OK) {
+    return PDM_ERROR;
+  }
+
+  return PDM_OK;
+}
+
+/**
+ *  @brief Stop recording
+  * @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_stop(void)
+{
+  if (HAL_SAI_DMAStop(&hsai_BlockB1) != HAL_OK) {
+    return PDM_ERROR;
+  }
+
+  return PDM_OK;
+}
+
+/**
+ *  @brief Pause recording
+  * @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_pause(void)
+{
+  if (HAL_SAI_DMAPause(&hsai_BlockB1) != HAL_OK) {
+    return PDM_ERROR;
+  }
+
+  return PDM_OK;
+}
+
+/**
+ *  @brief Resume recording
+  * @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_resume(void)
+{
+  if (HAL_SAI_DMAResume(&hsai_BlockB1) != HAL_OK) {
+    return PDM_ERROR;
+  }
+
+  return PDM_OK;
+}
+
+/**
+ *  @brief Half Callback
+ *  @param pBuf: pointer to a buffer to store data
+ *  @retval None
+ */
+void pdm_low_level_copy_half_cplt(uint16_t *pBuf)
+{
+  uint32_t index;
+  uint16_t *DataTempSAI = SAI_InternalBuffer;
+
+  for (index = 0; index < (SAI_InternalBufferSize / 4U); index++) {
+    pBuf[index] = (DataTempSAI[index]);
+  }
+}
+
+/**
+ *  @brief Complet Callback
+ *  @param pBuf: pointer to a buffer to store data
+ *  @retval None
+ */
+void pdm_low_level_copy_cplt(uint16_t *pBuf)
+{
+  uint32_t index;
+  uint16_t *DataTempSAI = &SAI_InternalBuffer[SAI_InternalBufferSize / 4U];
+
+  for (index = 0; index < (SAI_InternalBufferSize / 4U); index++) {
+    pBuf[index] = (DataTempSAI[index]);
+  }
+}
+
+/* This function is defined weak in the Arduino Core, so we can override it */
+
+/**
+ *  @brief System Clock Parameters Configuration
+  * @param None
+  * @retval None
+ */
+void SystemClock_Config(void)
+{
+  RCC_OscInitTypeDef RCC_OscInitStruct = {0};
+  RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};
+  RCC_PeriphCLKInitTypeDef PeriphClkInit = {0};
```

Comment:
Hi @sarachiodo ,
I noticed that the Arduino compiler gives a warning. We need to replace {0} with simply {}